[Intel MKL] Requantization op perchannel support

nammbash opened the pull request:
## Requantization_Perchannel_op

1. Adding two new hidden operators to support requantization to be performed perchannel and provided the mkl implementations for the same.

        1. requantize_per_channel_op
        2. requantization_range_per_channel_op

2. Provided Unit test to test the implementation logic of both these above mentioned ops.

nammbash committed:
requantization_op_perchannel_support

nammbash committed:
requantization_op_perchannel_support

nammbash committed:
requantization_op_perchannel_support

nammbash committed:
Merge branch 'requantization_op_perchannel_support' of https://github.com/Intel-tensorflow/tensorflow into requantization_op_perchannel_support

nammbash committed:
Merge branch 'requantization_op_perchannel_support' of https://github.com/Intel-tensorflow/tensorflow into requantization_op_perchannel_support

nammbash commented:
@hgadig 
Reopening this PR: https://github.com/tensorflow/tensorflow/pull/25038 which had unusual git rebasing problem.

nammbash committed:
fix api_test to succeed in Eigen along with config=mkl option

nammbash commented:
@penpornk can you review this PR and provide updates or marge please!

penpornk reviewed (changes requested):
Thank you for the PR! I need a clarification before I can review the test cases further, mainly about the range of the convolved quantized tensors,
```c++
// The Range of the Convolved T*W1 is 2^31 * 5.0/255 * 2.0/127 = 663110.59
```
Why is it `2^31` and not `(2^31-1)`? I have a separate comment for this. Please reply there. Thank you!

nammbash committed:
fix formatting manually for pbtxt and ignoring clang formatting

nammbash commented on tensorflow/core/kernels/mkl_requantization_range_per_channel_op.cc:
```diff
@@ -0,0 +1,110 @@
+    bool is_non_negative = true;
+    Eigen::array<int, 2> shuffling({1, 0});
+    auto input_matrix = input.flat_inner_dims<qint32>();
+    auto transposed_input = input_matrix.shuffle(shuffling);
```
For better cache efficiency.

nammbash replied:
Not sure about the Eigen implementation but the mkl version of transpose was fast. Another point to note is that this op is a calibration step, meaning it will be changed to a const node for the final inference graph.

nevertheless ToDo and comments are due and provided.

penpornk replied:
Got it. Appreciate the TODO and the note about const node!

nammbash committed:
Review changes

nammbash commented on tensorflow/core/api_def/base_api/api_def_RequantizationRangePerChannel.pbtxt:
```diff
@@ -0,0 +1,4 @@
+op {
+graph_op_name:
+  "RequantizationRangePerChannel" visibility : HIDDEN
+}
```
Clang was run on these files which mangled the format. Should be looking good now

nammbash replied:
Clang was run on these files which mangled the format. Should be looking good now

nammbash replied:
Done!

nammbash commented on tensorflow/core/kernels/BUILD:
```diff
@@ -6981,6 +6983,67 @@ tf_mkl_kernel_library(
+            "@gemmlowp",
+            ":transpose_functor",
+            "@mkl_dnn",
+            "//third_party/mkl:intel_binary_blob",
```
Done!

nammbash replied:
Done!

nammbash commented on tensorflow/core/kernels/mkl_requantization_range_per_channel_op.cc:
```diff
@@ -0,0 +1,110 @@
+
+    float out_min_max = std::numeric_limits<float>::min();
+    for (size_t i = 0; i < depth; i++) {
+      if (out_min_max < ranges[i]) out_min_max = ranges[i];
```
The previous loop is parallelised using Open MP threads. Hence we obtain the final numbers after that is done. Updated comments to explain this further.

penpornk replied:
I think you forgot to delete the old code and uncomment the new code?

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  // The output Tensor T is in int32 whose range is [-2^31, 2^31]
+  // The Range of the Convolved T*W1 is 2^31 * 5.0/255 * 2.0/127 = 663110.59
+  // So Range of Convolved T*W1 in int32[-2^31, 22^31] that maps to
+  // orig T Range in FP32[0,5.0] * [-2.0, 2.0] is [-663110.59, 663110.59]
```
That is correct! It should be (2^31-1). Updated the comment although the perturbation is very small that it doesnot affect the value in question in this test.

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+                            input_tensor_qint32.flat<qint32>());
+
+  // Calculate the Min and max from the ranges
+  float ch1_min = -1.0 * range_weights_ch1;
```
:) for extra clarity during initializing constant values and sometimes for casting it to float. In this case the variable is also a float so not necessary! Changed them nevertheless

nammbash commented on tensorflow/core/kernels/BUILD:
```diff
@@ -6104,7 +6104,9 @@ tf_kernel_library(
     size = "small",
-    srcs = ["requantization_range_op_test.cc"],
+    srcs = [
+        "requantization_range_op_test.cc",
```
Done!

nammbash commented on tensorflow/core/kernels/BUILD:
```diff
@@ -6981,6 +6983,67 @@ tf_mkl_kernel_library(
+        "no_op.h",
+        "reference_gemm.h",
+    ],
+    deps = if_mkl(
```
Done!

nammbash commented on tensorflow/core/kernels/BUILD:
```diff
@@ -6981,6 +6983,67 @@ tf_mkl_kernel_library(
+            "//tensorflow/core:nn_ops_op_lib",
+            "//third_party/eigen3",
+            "@gemmlowp",
+            ":transpose_functor",
```
Done!

nammbash commented:
@rthadur addressed all review comments!

penpornk commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  range_weights_ch1 = 663110.59;  // For W1 channel
+  range_weights_ch2 = 994665.88;  // For W2 Channel
+
+  // We Fill the inputs Tensor T qint32 with arbitrary int32 values
```
```suggestion
  // We fill the input Tensor T qint32 with arbitrary int32 values.
```

nammbash replied:
done!

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+
+TEST_F(MklRequantizatedOpsTest, RequantizePerChannelTest_Basic) {
+  // Let us setup the tensor and inputs before we run this op.
+  float clip_max_value = pow(2, 31);
```
done!

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+
+TEST_F(MklRequantizatedOpsTest, RequantizationRangePerChannelTest_ClipMax) {
+  // Let us setup the tensor and inputs before we run this op.
+  float clip_max_value = 6;  // Can be used as 6 for Relu 6 activations.
```
I purposefully had it to be different from the actual Tensor input name. Nevertheless changed!

penpornk replied:
Got it. And thank you for changing! Using a different word made me wonder if the two names do the same thing (until I saw the code that confirmed so), so I'm glad that you changed it.

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+                            input_tensor_qint32.flat<qint32>());
+
+  // Calculate the Min and max from the ranges
+  float ch1_min = -1.0 * range_weights_ch1;
```
done!

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  // Calculate the Min and max from the ranges
+  float ch1_min = -1.0 * range_weights_ch1;
+  float ch1_max = range_weights_ch1;
+  float ch2_min = -1.0 * range_weights_ch2;
```
done!

nammbash commented on tensorflow/core/kernels/mkl_requantization_range_per_channel_op.cc:
```diff
@@ -0,0 +1,110 @@
+    const Tensor& input_min = ctx->input(kInputMin);
+    const Tensor& input_max = ctx->input(kInputMax);
+
+    size_t depth = input_max.NumElements();
```
done!

nammbash commented on tensorflow/core/kernels/mkl_requantization_range_per_channel_op.cc:
```diff
@@ -0,0 +1,110 @@
+      int32_t max_per_channel = max();
+      int32_t abs_max =
+          std::max(std::abs(min_per_channel), std::abs(max_per_channel));
+      float scale =
```
done!

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  // Step 3: Assumption of quantizedconv on quantized input&weights(per channel)
+  // ---------------------------------------------------------------------------
+  // The input T and weights W1 (or W2) will be convolved.
+  // The output Tensor T is in int32 whose range is [-(2^31-1), 2^31-1)
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  // ---------------------------------------------------------------------------
+  // The input T and weights W1 (or W2) will be convolved.
+  // The output Tensor T is in int32 whose range is [-(2^31-1), 2^31-1)
+  // The Range of the convolved T*W1 is ((2^31)-1) * 5.0/255 * 2.0/127 = 663110.59
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  // The input T and weights W1 (or W2) will be convolved.
+  // The output Tensor T is in int32 whose range is [-(2^31-1), 2^31-1)
+  // The Range of the convolved T*W1 is ((2^31)-1) * 5.0/255 * 2.0/127 = 663110.59
+  // So the range of convolved T*W1 in int32[-(2^31-1), 22^31-1] that maps to
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  // The output Tensor T is in int32 whose range is [-(2^31-1), 2^31-1)
+  // The Range of the convolved T*W1 is ((2^31)-1) * 5.0/255 * 2.0/127 = 663110.59
+  // So the range of convolved T*W1 in int32[-(2^31-1), 22^31-1] that maps to
+  // orig T Range in FP32[0,5.0] * [-2.0, 2.0] is [-663110.59, 663110.59]
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  // So the range of convolved T*W1 in int32[-(2^31-1), 22^31-1] that maps to
+  // orig T Range in FP32[0,5.0] * [-2.0, 2.0] is [-663110.59, 663110.59]
+
+  // The Range of the convolved T*W2 is 2^31-1 * 5.0/255 * 3.0/127 = 994665.88
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  // orig T Range in FP32[0,5.0] * [-2.0, 2.0] is [-663110.59, 663110.59]
+
+  // The Range of the convolved T*W2 is 2^31-1 * 5.0/255 * 3.0/127 = 994665.88
+  // So Range of convolved T*W2 in int32[-(2^31-1), 22^31-1] that maps to
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+
+  // The Range of the convolved T*W2 is 2^31-1 * 5.0/255 * 3.0/127 = 994665.88
+  // So Range of convolved T*W2 in int32[-(2^31-1), 22^31-1] that maps to
+  // orig T Range in FP32 [0, 5.0] * [-3.0, 3.0]  is [-994665.88, 994665.88]
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+
+  // Step 4: Assumption output above is fed to Requantization_range_perchannel
+  // --------------------------------------------------------------------------
+  // Here we recalculate the new Range for convolved T*W so that we
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+
+  // Step 5: Define and run requantization_range_perchannel
+  // -------------------------------------------------------
+  // Define, Create and initalize the OP in question.
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+// of the weights is calculated per channel.
+TEST_F(MklRequantizatedOpsTest, RequantizationRangePerChannelTest_Basic) {
+  // Let us set up the tensor and inputs before we run this op.
+  float clip_max_value = pow(2, 31);
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_per_channel_op.cc:
```diff
@@ -0,0 +1,170 @@
+        float min_max_from_vec = std::max(std::abs(input_min_vec_data[i]),
+                                          std::abs(input_max_vec_data[i]));
+        scales[i] =
+            factor * (min_max_from_vec / requested_min_max / (float)(1L << 31));
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+
+  // Step 5: Define and run requantization_range_perchannel
+  // -------------------------------------------------------
+  // Define, Create and initalize the OP in question.
```
done

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+
+  // Step 7: Define and run requantize_perchannel
+  // --------------------------------------------
+  // The Output of Requantization_range_op_per_channel which calculated the
```
done

nammbash committed:
review comments 2a

penpornk committed:
Update tensorflow/core/kernels/mkl_requantize_ops_test.cc

Co-Authored-By: nammbash <38869685+nammbash@users.noreply.github.com>

penpornk committed:
Update tensorflow/core/kernels/mkl_requantize_ops_test.cc

Co-Authored-By: nammbash <38869685+nammbash@users.noreply.github.com>

penpornk committed:
Update tensorflow/core/kernels/mkl_requantize_ops_test.cc

Co-Authored-By: nammbash <38869685+nammbash@users.noreply.github.com>

penpornk committed:
Update tensorflow/core/kernels/mkl_requantize_ops_test.cc

Co-Authored-By: nammbash <38869685+nammbash@users.noreply.github.com>

penpornk commented:
Looks like the other author that @googlebot has a CLA problem with is me. Manually setting CLA to yes then.

penpornk commented on tensorflow/core/api_def/base_api/api_def_RequantizePerChannel.pbtxt:
```diff
@@ -1,4 +1,66 @@
+    name: "out_type"
+    description: <<END
+The quantized type of output tensor that needs to be converted. 
+input tensor.
```
This seems like a stray line?

nammbash replied:
yes! removed

penpornk commented on tensorflow/core/kernels/mkl_requantization_range_per_channel_op.cc:
```diff
@@ -0,0 +1,125 @@
+    auto input_matrix = input.flat_inner_dims<qint32>();
+
+    // TODO: verify performance of not transposing and finding the min max directly
+    // from input_matrix vs the one presentd below of transposing and using the
```
```suggestion
    // from input_matrix vs the one presented below of transposing and using the
```

nammbash replied:
done!

penpornk commented on tensorflow/core/kernels/mkl_requantization_range_per_channel_op.cc:
```diff
@@ -0,0 +1,125 @@
+
+        // thread-local out_min_max. 
+        //if (out_min_max < ranges[i]) out_min_max = ranges[i];
+      } 
```
The indentation-level of this bracket seems wrong.

penpornk commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  // orig T range in FP32[0, 5.0] * [-2.0, 2.0] is [-663110.59, 663110.59]
+
+  // The range of the convolved T*W2 is (2^31-1) * 5.0/255 * 3.0/127 = 994665.88
+  // So the range of convolved T*W1 in int32(-2^31, 2^31) that maps to
```
```suggestion
  // So the range of convolved T*W2 in int32(-2^31, 2^31) that maps to
```

nammbash replied:
yup

penpornk reviewed (changes requested):
Thank you for the edits! I think there are some editing mistakes.

nammbash commented:
@penpornk haven't completed it! will ping you late this evening or tomorrow and confirm edits. :)

nammbash committed:
omp reduction feature and clang formatting

penpornk commented:
@nammbash Got it. Please take your time. :) And don't worry about the CLA. I have a feeling google bot will do this a lot. I'll just change it again after we are done with the PR.

nammbash commented:
https://github.com/tensorflow/tensorflow/pull/25203#discussion_r252807962
good suggestion. Learnt something new today thank you!
 although it is reduction(max : out_min_max) and worked only with the max function and not without it! Changes done.

nammbash committed:
change comments for test

nammbash commented on tensorflow/core/kernels/mkl_requantization_range_per_channel_op.cc:
```diff
@@ -0,0 +1,112 @@
+  float clip_value_max_ = std::numeric_limits<float>::infinity();
+  const int kInputTensorIndex = 0;
+  const int kInputMinIndex = 1;
+  const int kInputMaxIndexIndex = 2;
```
done!

nammbash committed:
change comments for test 2

nammbash commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,297 @@
+  // --------------------------------------------
+  // The Output of Requantization_range_op_per_channel which calculated the
+  // new ranges of int8 is fed to the requantize per channel op.
+  // Here the Values of Convolved T*W is converted from int32 to int8.
```
Done!

nammbash commented on tensorflow/core/api_def/base_api/api_def_RequantizationRangePerChannel.pbtxt:
```diff
@@ -0,0 +1,48 @@
+  attr {
+    name: "clip_value_max"
+    description: <<END
+The maximum value of the output that needs to to be clipped.
```
done!

shahzadlone commented on tensorflow/core/kernels/mkl_requantize_per_channel_op.cc:
```diff
@@ -0,0 +1,170 @@
+      if (out_type_ == DT_QINT8) DCHECK(input_requested_min_float < 0.0f);
+
+      const float factor = (out_type_ == DT_QINT8) ? 127.0f : 255.0f;
+      float requested_min_max = std::max(std::abs(input_requested_min_float),
```
is this also suppose to be ```const```?

nammbash committed:
review comments changes

penpornk commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,300 @@
+  EXPECT_NEAR(-14.8217, output_min, 0.002);
+  EXPECT_NEAR(14.8217, output_max, 0.002);
+
+  // output range is made use in requantizePerChannelTest_Basic
```
```suggestion
  // Output range is made use in RequantizePerChannelTest_Basic
```
The test name is a name so let's keep the beginning capitalized.

nammbash replied:
Sure!

penpornk commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,300 @@
+  // The output tensor T is in int32 whose range is [-2^31, 2^31).
+  // For simplicity, we truncate the range to (-2^31, 2^31) to make it
+  // symmetric.
+  // The range of the convolved T*W1 is ((2^31)-1) * 5.0/255 * 2.0/127 =
```
Could you please merge this line with the line above so the two sentences are in three instead of four lines? Maybe like this (or some other ways you like)
```c++
  // For simplicity, we truncate the range to (-2^31, 2^31) to make it
  // symmetric. The range of the convolved T*W1 is 
  // ((2^31)-1) * 5.0/255 * 2.0/127 = 663110.59
```
(Sorry for nitpicking. I just don't like orphan words.)

nammbash replied:
No worries on suggestions were all good and keep at it. :)  I too do not like orphan words and even numbers. The comments had both. Effects of internal CI process and clang. Should be reformatted now.

penpornk commented on tensorflow/core/kernels/mkl_requantization_range_per_channel_op.cc:
```diff
@@ -0,0 +1,124 @@
+          scale * static_cast<float>(abs_max) / static_cast<float>(1L << 31);
+      if (min_per_channel < 0) is_non_negative = false;
+
+      // thread-local out_min_max.
```
```suggestion
      // Thread-local out_min_max.
```

nammbash replied:
done!

penpornk reviewed (changes requested):
Thanks so much for all your careful edits (and for your patience)! I have a few nitpicks and then we should be good to go. :)

Transpose: Thanks for the explanation!
Reduction: You're welcome! And you are right that it should be max reduction. For some reason I thought we were doing `min` when I typed that. Sorry about that!

nammbash committed:
some small changes

nammbash committed:
some small changes

nammbash commented:
@penpornk changes done!
side note again: there is another PR that needs to go in related to perchannel. (#25506)

penpornk commented on tensorflow/core/kernels/mkl_requantize_per_channel_op.cc:
```diff
@@ -0,0 +1,173 @@
+
+// See docs in ../ops/array_ops.cc.
+
+#define EIGEN_USE_THREADS
```
Let's move this below `#ifdef INTEL_MKL`.

penpornk commented on tensorflow/core/kernels/requantize.cc:
```diff
@@ -100,4 +100,10 @@ REGISTER_KERNEL_BUILDER(Name("Requantize")
                             .TypeConstraint<quint8>("out_type"),
                         RequantizeOp<qint32, quint8>);
 
+REGISTER_KERNEL_BUILDER(Name("Requantize")
```
This seems redundant with the registration before it?

penpornk commented on tensorflow/core/ops/math_ops.cc:
```diff
@@ -1688,6 +1688,45 @@ inputs: Must all be the same size and shape.
 
 #endif  // INTEL_MKL
```
The registration of `RequantizePerChannel` and `RequantizationRangePerChannel` ops must be inside the `#ifdef INTEL_MKL` guard, because their kernel builders are only defined with config=mkl.

nammbash replied:
The Op needs to be available outside of MKL also, as there is no TF(Eigen) Op available with the same name. This is required for pbtxt , else API tests will fail. let me know if this is not clear. Tomorrow in the meeting with Intel and Google AG Ramesh might add more or else I can explain in detail.

nammbash replied:
these 2 tests will pass only the way I have it now for both Eigen and MKL
1.//tensorflow/core/kernels:mkl_requantize_ops_test
2.tensorflow/core/api_def/api_test

penpornk replied:
Thank you for the explanation! I was under the impression that registering the op without having a kernel builder (in the non-MKL case) would have a problem. I agree that it works fine. Saw that a lot of other MKL code does it this way too. :)

penpornk commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,298 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
```
While we are at it, let's also guard the tests. Please add `if defined(INTEL_MKL) && defined(ENABLE_MKL)`.

penpornk commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,298 @@
+  EXPECT_NEAR(range_op_output_max, output_max, 0.002);
+}
+
+}  // namespace tensorflow
```
Please add `#endif  // INTEL_MKL && ENABLE_MKL`.

penpornk reviewed (changes requested):
Another MKL PR broke a regular TF build because it registers MKL-only ops without the `INTEL_MKL` guard. So I did a little once-over and found a few more things. Sorry for the inconvenience!

nammbash committed:
MKL conditional compilation

nammbash committed:
cleanup requantize

penpornk commented on tensorflow/core/api_def/base_api/api_def_RequantizePerChannel.pbtxt:
```diff
@@ -0,0 +1,65 @@
+op {
+graph_op_name: "RequantizePerChannel"
```
There should be indentation here. Did you use the right formatter?

nammbash replied:
No worries! I am not trusting formatters now! Its on me and sorry about that. Will be easier next time around with my PRs.

penpornk reviewed (changes requested):
I'm sorry I have to ask for changes again, but the formatting for the pbtxt files doesn't look right. The first few lines in `op` need indentation.

nammbash committed:
indentation in pbtxt needs to be corrected

penpornk reviewed (dismissed):
@nammbash No worries! Thanks again for your PR!
Also, thanks for your comments, @shahzadlone !

rthadur commented:
@nammbash gentle ping to  check test failures.

nammbash commented:
@rthadur Interesting! 
1. some of these are not related to CPU. (Like the GPU CC) . Correct me if I am wrong.
2. the test //tensorflow/tools/api/tests:api_compatibility_test passes in c++. Is there a python version of the same test? Let me check into that

nammbash commented:
@rthadur @penpornk 
Interesting findings on the build server failures. per your issue-comment: https://github.com/tensorflow/tensorflow/pull/25203#issuecomment-461208892

**I am ignoring the following failures as it does not pertain to my PR:**
```
GPU CC
GPU python 3
MacOS
Windows Bazel GPU
```
**The pertaining ones then is this**

```Ubuntu python2```

**google build server error as displayed by web link points to the test:**
```
tensorflow/tools/api/tests:api_compatibility_test

and hints problem to be in the pbtxt apis of RequantizePerchannel. RequantizeRnagePerchannel and requantize.
```
**Findings when I run it locally**
1. Bazeltest  succeeds if I use "bazel test" directly. 
```
bazel --output_base=../google_fail test tensorflow/tools/api/tests:api_compatibility_test
```
2. if I build it separately and then run it as the log online seems to indicate
```
bazel --output_base=../google_fail build tensorflow/tools/api/tests:api_compatibility_test
then
a. bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True (success)
b. bazel-bin/tensorflow/tools/api/tests/api_compatibility_test  (Failed)

yet the failure seems to not have any mention of perchannel or requantize ops.
```
**Analysis Conclusion:(see findings below)**
```
1.  I believe that google build server is having some issues, 
or 
2. maybe with this specific run 
or 
3. bazel bug??! Internal bazel build is different than what I have locally?? locally 
I am using bazel version: 0.19.2 running on Ubuntu xfce: 4.12
```
**Hence I believe the failures are not related to the PR and related to bazel itself? correct me if this seems not the case in any of your internal testing.**

penpornk commented:
@nammbash You are right about other failures are unrelated. The API one is indeed because of this PR. 

I forgot about updating the API golden files. Sorry! Basically, when you add any new API pbtxt, you need to add an entry to the golden files too. I believe running the test with `--update_golden` modifies the necessary goldenfiles (e.g., tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt, ensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt, etc) for you.

For example, see how the golden files are modified in https://github.com/tensorflow/tensorflow/commit/de87e628e6d89382783ea948c21fa66b182d69d3. (You shouldn't modify it yourself. Let the test modify them and add the changes to the PR.)

nammbash committed:
additional golden files for fixing

nammbash commented:
@penpornk modifications done and committed. Please review and merge.
@rthadur  FYI

Action:
build test separately and then run
```
bazel --output_base=../test_folder build tensorflow/tools/api/tests:api_compatibility_test
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test  (Failed)

but

bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens True 
Fails and updated the files

bazel --output_base=../test_folder build tensorflow/tools/api/tests:api_compatibility_test
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test  (Success)

rave991 commented:
Citrajac@gmail.com

penpornk commented:
@nammbash Hmm. This is weird. The golden files aren't related to your PR. (But the test passed without your PR.) I'll ask someone who actually knows more about them. Will get back to you soon.

nammbash commented:
@penpornk 
:) Exactly my point from my comments yesterday: https://github.com/tensorflow/tensorflow/pull/25203#issuecomment-461262408.

thank you for the internal reviews!

penpornk commented:
@nammbash So @annarev told me we actually don't need to modify the golden files at all since the visibility is hidden. The mismatch on your part might be the estimator version mismatch. But then I'm not sure why our Kokoro test failed on your PR and not on the current repo. Could you please try syncing your local copy? (Also, please drop the last commit changing the golden files.) Sorry for the inconvenience!

penpornk commented:
@nammbash I'm looking at the test results from [last night](https://source.cloud.google.com/results/invocations/dcc891bd-2e73-4b95-9a72-32de274bd127/log). The differences were actually the entries for your ops (`RequantizationPerChannel` and `RequantizationRangePerChannel`). That's why I thought they needed to be added to the golden files. Now there two questions: why did you see different results, and why did the test try to add your 'hidden' ops to the golden files. 

```
E0207 07:22:46.303906 140336513210112 api_compatibility_test.py:216] TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.
If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.
    $ bazel build tensorflow/tools/api/tests:api_compatibility_test
    $ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.
E0207 07:22:46.304153 140336513210112 api_compatibility_test.py:217] 1 differences found between API and golden.
Issue 1	: None :
---
+++
@@ -2645,10 +2645,18 @@
     argspec: "args=[\'input\', \'input_min\', \'input_max\'], varargs=None, keywords=None, defaults=None"
   }
   member_method {
+    name: "RequantizationRangePerChannel"
+    argspec: "args=[\'input\', \'input_min\', \'input_max\', \'clip_value_max\'], varargs=None, keywords=None, defaults=None"
+  }
+  member_method {
     name: "Requantize"
     argspec: "args=[\'input\', \'input_min\', \'input_max\', \'requested_output_min\', \'requested_output_max\', \'out_type\'], varargs=None, keywords=None, defaults=None"
   }
   member_method {
+    name: "RequantizePerChannel"
+    argspec: "args=[\'input\', \'input_min\', \'input_max\', \'requested_output_min\', \'requested_output_max\', \'out_type\'], varargs=None, keywords=None, defaults=None"
+  }
+  member_method {
     name: "Reshape"
     argspec: "args=[\'tensor\', \'shape\'], varargs=None, keywords=None, defaults=None"
   }
```

nammbash commented:
@penpornk I was confused last afternoon on this question too. hence the question is there a build server/ specific run error.

Interestingly  bazel test passes but bazel build and then running test fails! Why?

bazel --output_base=../test_folder test tensorflow/tools/api/tests:api_compatibility_test passes.

nammbash committed:
Revert "additional golden files for fixing"

This reverts commit f47f02cec0ec05a823f5ac8b5fc8373d598d7f1a.

nammbash committed:
Merge branch 'master' into requantization_op_perchannel_support

nammbash commented:
@penpornk 
Status Quo: https://github.com/tensorflow/tensorflow/pull/25203#issuecomment-461573084

Q1: why did you see different results
A1: No more after the merge with master from yesterday. I see the same error.

Q2: why did the test try to add your 'hidden' ops to the golden files
This is confusing. test fails and investigating. Not sure why hidden is doing this. Yet still the changes made by the golden file is for the estimators and not my op. but the test passes after these changes? Is the problem even related! Weird!

annarev commented:
To add to @penpornk reply, there is this RFC for exporting all ops to the raw_ops namespace:
https://github.com/tensorflow/community/blob/master/rfcs/20181225-tf-raw-ops.md
This seems to include hidden ops as well. @alextp  would know more.

Also, to be more specific, --only_test_core_api=True needs to be passed in to api_compatibility_test when updating goldens. This flag is passed automatically when running test with "bazel test", but not when running it for golden update.

nammbash committed:
Merge branch 'master' into requantization_op_perchannel_support

alextp commented:
Anna, should we update the error message in the test to instruct people to
pass it when updating goldens?

On Thu, Feb 7, 2019 at 3:26 PM annarev <notifications@github.com> wrote:

> To add to @penpornk <https://github.com/penpornk> reply, there is this
> RFC for exporting all ops to the raw_ops namespace:
>
> https://github.com/tensorflow/community/blob/master/rfcs/20181225-tf-raw-ops.md
> This seems to include hidden ops as well. @alextp
> <https://github.com/alextp> would know more.
>
> Also, to be more specific, --only_test_core_api=True needs to be passed in
> to api_compatibility_test when updating goldens. This flag is passed
> automatically when running test with "bazel test", but not when running it
> for golden update.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tensorflow/tensorflow/pull/25203#issuecomment-461632527>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAATxcnOkmT6QfQhTfmMiMp6Wfpbytviks5vLLYwgaJpZM4aTxMo>
> .
>


-- 
 - Alex

alextp commented:
And yes, the raw_ops namespace includes hidden ops. Feel free to bypass API
review for it, though.

On Thu, Feb 7, 2019 at 4:14 PM Alexandre Passos <apassos@google.com> wrote:

> Anna, should we update the error message in the test to instruct people to
> pass it when updating goldens?
>
> On Thu, Feb 7, 2019 at 3:26 PM annarev <notifications@github.com> wrote:
>
>> To add to @penpornk <https://github.com/penpornk> reply, there is this
>> RFC for exporting all ops to the raw_ops namespace:
>>
>> https://github.com/tensorflow/community/blob/master/rfcs/20181225-tf-raw-ops.md
>> This seems to include hidden ops as well. @alextp
>> <https://github.com/alextp> would know more.
>>
>> Also, to be more specific, --only_test_core_api=True needs to be passed
>> in to api_compatibility_test when updating goldens. This flag is passed
>> automatically when running test with "bazel test", but not when running it
>> for golden update.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/tensorflow/tensorflow/pull/25203#issuecomment-461632527>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAATxcnOkmT6QfQhTfmMiMp6Wfpbytviks5vLLYwgaJpZM4aTxMo>
>> .
>>
>
>
> --
>  - Alex
>


-- 
 - Alex

annarev commented:
@alextp I think we could just change the default value, so that passing this flag is not required.

nammbash committed:
adding new ops api to the new shared raw ops file in tensorflow/tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt

penpornk reviewed (approved):
Looks good! Thank you!
I'll skip the `api review` tag as @alextp suggested.

nammbash commented:
@penpornk and @annarev thank you for the comments.
Looks like that was a massive overhaul change with the new API and seems like I was the first to get caught trying the new method amidst @alextp changes! :)

What was done:
1. Added the 2 new ops in file tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt
Looks like alphabetical order matters too. (So some automated method to add them / generate them would be beneficial in future)
2. Ran the test: 
bazel --output_base=../test test --cache_test_results=no tensorflow/tools/api/tests:api_compatibility_test
Success.
3. Separate passes only with the flag as mentioned.
bazel --output_base=../test build --cache_test_results=no tensorflow/tools/api/tests:api_compatibility_test
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --only_test_core_api=True

@penpornk and @annarev and @alextp thank you all for the help!
Suggestions:
It is confusing to have two tests for api.(whcih shoudl i worry about?)
a. tensorflow/tools/api/tests:api_compatibility_test
and
b.tensorflow/core/api_def/api_test

2. Yes some error message on what needs to be done pointing at the api file would be helpful, including mention of alphabetical order(maybe)!

annarev commented:
1.
@nammbash the two tests are different and both important:
a. tensorflow/tools/api/tests:api_compatibility_test: compares TensorFlow Python public API to golden files. It is important for checking backwards compatibility and monitoring API changes.
b. tensorflow/core/api_def/api_test: checks consistency of api_def_OpName.pbtxt files specifically. For e.g. it checks that all registered ops should have api_def_OpName.pbtxt file. These .pbtxt files are used for all language APIs, so this is not specific to Python unlike api_compatibility_test.

2. We should definitely look at improving error messages. Ideally, golden files should be updated automatically when running api_compatibility_test with --update_goldens=True. But I can see that at times this approach could hit issues (for e.g. in this case, you needed to pass additional flag).

penpornk commented on tensorflow/core/kernels/BUILD:
```diff
@@ -6986,6 +6986,66 @@ tf_mkl_kernel_library(
+    name = "mkl_requantize_ops_test",
+    size = "small",
+    srcs = ["mkl_requantize_ops_test.cc"],
+    deps = [
```
Please add 
```
    tags = ["no_mac"],  // TODO(penporn): Re-enable the test on MacOS.
```
 before this line.

nammbash replied:
Done! FYI

nammbash commented:
@rthadur I think the failures are all not related to my PR?

penpornk reviewed (changes requested):
**Ubuntu Sanity:** Please run `buildifier tensorflow/core/kernels/BUILD`

**MacOS Python2 and CC:** The error message is: 
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //tensorflow/core/kernels:mkl_requantize_ops_test
-----------------------------------------------------------------------------
dyld: malformed mach-o: load commands size (33344) > 32768
Abort trap: 6
```
`dyld` is Mac's dynamic linker. I have seen this error on MacOS with `mkl_conv_ops_test` for a while and am finding out why. For now let's just disable your test on MacOS. 

**Windows Bazel GPU:** Unrelated to your PR.
**Ubuntu Python 3 PIP:** Unrelated to your PR.

nammbash committed:
build file buildifier formatter

nammbash commented:
@penpornk  thank you!
```
FYI
tags = ["no_mac"],  # TODO(penporn): Re-enable the test on MacOS.
```
please re run the tests.

penpornk reviewed (approved):
Thank you!

nammbash commented:
@rthadur Who pulls and merges this PR?

hgadig commented:
> @rthadur can you please merge this PR?(making sure it is in before any new conflicts) thank you! :)

Sure. We are looking into this. As "rthadur" is off, i'll be taking this.

penpornk commented:
@nammbash Our internal test says the v1 golden file should be updated as well. Sorry that this wasn't reflected in our external tests. Did you see any changes to `tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt` when you updated the golden files? If so, please include it. If not, we need to figure out why.

nammbash commented:
@penpornk I updated tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt manually!
Didn't know there was a flag to do this?
--only_test_core_api=True --wil not modify anything.

are you expecting --update_goldens=True to do something?

penpornk reviewed (changes requested):
@nammbash The file that you had committed yesterday was v2. We also need v1. Yes, we expect `--update_goldens True` to update all the necessary golden files for you. The following commands should be all you need:
```
bazel build //tensorflow/tools/api/tests/api_compatibility_test
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --only_test_core_api=True --update_goldens=True
```
After this, you should see the modifications to `tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt` as well. Please include that in the PR.

penpornk commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,300 @@
+
+  // Step 5: Define and run requantization_range_perchannel
+  // -------------------------------------------------------
+  // Define, create and initalize the op in question.
```
```suggestion
  // Define, create and initialize the op in question.
```

penpornk commented on tensorflow/core/kernels/mkl_requantize_ops_test.cc:
```diff
@@ -0,0 +1,300 @@
+ public:
+  void Setup(Tensor &input_tensor_qint32, float &range_weights_ch1,
+             float &range_weights_ch2);
+  void TestBody(){};
```
```suggestion
  void TestBody(){}
```

penpornk commented on tensorflow/core/kernels/mkl_requantize_per_channel_op.cc:
```diff
@@ -0,0 +1,173 @@
+
+      output_min->flat<float>()(0) = input_requested_min_float;
+      output_max->flat<float>()(0) = input_requested_max_float;
+
```
Please remove this blank line.

penpornk reviewed (changes requested):
Also a few more edits, please.

penpornk committed:
Update tensorflow/core/kernels/mkl_requantize_ops_test.cc

Co-Authored-By: nammbash <38869685+nammbash@users.noreply.github.com>

penpornk committed:
Update tensorflow/core/kernels/mkl_requantize_ops_test.cc

Co-Authored-By: nammbash <38869685+nammbash@users.noreply.github.com>

nammbash committed:
remove balnk line

penpornk commented:
@nammbash Are you able to generate the changes to `tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt`? If not, here is the diff. I guess we can modify it directly this time. (Line number might be slightly different.)
```
@@ -2645,10 +2645,18 @@
     argspec: "args=[\'input\', \'input_min\', \'input_max\'], varargs=None, keywords=None, defaults=None"
   }
   member_method {
+    name: "RequantizationRangePerChannel"
+    argspec: "args=[\'input\', \'input_min\', \'input_max\', \'clip_value_max\'], varargs=None, keywords=None, defaults=None"
+  }
+  member_method {
     name: "Requantize"
     argspec: "args=[\'input\', \'input_min\', \'input_max\', \'requested_output_min\', \'requested_output_max\', \'out_type\'], varargs=None, keywords=None, defaults=None"
   }
   member_method {
+    name: "RequantizePerChannel"
+    argspec: "args=[\'input\', \'input_min\', \'input_max\', \'requested_output_min\', \'requested_output_max\', \'out_type\'], varargs=None, keywords=None, defaults=None"
+  }
+  member_method {
     name: "Reshape"
     argspec: "args=[\'tensor\', \'shape\'], varargs=None, keywords=None, defaults=None"
   }
```

hgadig commented:
@nammbash  Please sign the CLA in order to proceed with merging.

penpornk commented:
@hgadig He already signed the CLA. The bot detected his CLA and assigned `CLA: yes` since the beginning of the PR. Even the message right now says the bot found the CLA for @nammbash. What makes the bot assign `CLA: no` is because some of the commits are his accepting my online one-line suggestions (GitHub's new feature). For those commits, the author's emails that GitHub automatically used are `38869685+nammbash@users.noreply.github.com` and `38085909+penpornk@users.noreply.github.com`, which is not recognized by the googlebot. 

Googlebot wants us to verify that all the coauthors in this PR have signed the CLA. There are only two people contributing to this PR: @nammbash and I. So I think we are fine. :)

penpornk commented:
A quick note that I'm not setting CLA to yes yet until we are done with all the changes (since googlebot will keep setting it to no).

nammbash committed:
Added the new ops to the golden v1

nammbash commented:
@penpornk Sorry was in a meeting and hence the delay! Thank you for the code suggestions again, appreciate it! 

https://github.com/tensorflow/tensorflow/pull/25203#issuecomment-461952804 Here I meant to say v1 in the bottom comment file. Sorry about that confusion.(edited it)

Clearly there is difference between what google expect from these automation scripts and test flags and what I am experiencing locally. feels like I am beta testing the changes . :)

Anyways here is what I am experiencing from the latest master as available to external customers. Hoping it will identify the bug/process automation misses that can help google.
1. bazel test when I use the "test" option passes
```
bazel --output_base=../test2 test --cache_test_results=no //tensorflow/tools/api/tests:api_compatibility_test
```
2. Yet! bazel build + tests run separately have different behaviors
```
bazel --output_base=../test2 build --cache_test_results=no //tensorflow/tools/api/tests:api_compatibility_test 
```
then
trial 1. Failure. No changes to any Files!
``` bazel-bin/tensorflow/tools/api/tests/api_compatibility_test ```

trial 2.  Success (as it skipped 3 tests IMO) No changes in files
```
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --only_test_core_api=True
```

trial 3.   Success (as it skipped 3 tests IMO) No changes in files
``
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --only_test_core_api=True --update_goldens=True
```
trial 4. Success (as it skipped 3 tests IMO) - But 4 file changes(3 new and 1 addition)
```
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --update_goldens=True

files changed are as below:
(modified 3)
modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.-mode-keys.pbtxt
modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.experimental.pbtxt
modified:   tensorflow/tools/api/golden/v2/tensorflow.estimator.pbtxt
(added 1)
tensorflow/tools/api/golden/v2/tensorflow.estimator.inputs.pbtxt

Note: If I build with these changes in trial 4 , Trial 1 will pass
```
**NOTE: None of this Modifies the tensorflow/tools/api/golden/v1/tensorflow.raw_ops.pbtxt or it similar v2 part.**

Yesterday I changed the v2 manually. ( tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt ) 
Now I changed v1 manually ( tensorflow/tools/api/golden/v2/tensorflow.raw_ops.pbtxt)

Question: Do you want me to push all changes(modified files from Trial 4 above?)

penpornk reviewed (approved):
@nammbash Thank you for modifying the file manually! Your Trial 3 shouldn't have skipped three tests. If it hadn't skipped the tests, the `v1` file would have been modified too. We don't need estimator-related files from Trial 4. We should be good now, I think.

@annarev I tried fetching his PR and could replicate his results. The v1 test was skipped even when `tensorflow/tools/api/golden/v1/tensorflow.estimator.inputs.pbtxt` was in the branch. This is what I did
```
$ git fetch origin pull/25203/head:test
$ git checkout test
$ git pull origin master
$ bazel build //tensorflow/tools/api/tests:api_compatibility_test
$ bazel-bin/tensorflow/tools/api/tests/api_compatibility_test --only_test_core_api=True --update_goldens=True
Running tests under Python 2.7.13: /usr/bin/python
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibility
[  SKIPPED ] ApiCompatibilityTest.testAPIBackwardsCompatibility
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1
[  SKIPPED ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV1
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
I0210 22:35:59.131952 140640498542336 api_compatibility_test.py:243] No differences found between API and golden.
[       OK ] ApiCompatibilityTest.testAPIBackwardsCompatibilityV2
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessage
W0210 22:35:59.147357 140640498542336 deprecation.py:323] From /usr/local/google/home/penporn/software/tensorflow/bazel-bin/tensorflow/tools/api/tests/api_compatibility_test.runfiles/org_tensorflow/tensorflow/python/util/decorator_utils.py:145: VARIABLES (from tensorflow.python.framework.ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.GraphKeys.GLOBAL_VARIABLES` instead.
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessage
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV1
[ RUN      ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[       OK ] ApiCompatibilityTest.testNoSubclassOfMessageV2
[ RUN      ] ApiCompatibilityTest.test_session
[  SKIPPED ] ApiCompatibilityTest.test_session
----------------------------------------------------------------------
Ran 7 tests in 1.824s

OK (skipped=3)
```

nammbash commented:
@penpornk @annarev thank you! Let me know if anything else is required from me to push this PR!

nammbash commented:
@hgadig the 1 failure: Ubuntu Python3 PIP — Internal CI build failed  is not related to my changes in this PR. Let me know if anything else is required my side to merge this PR!

penpornk commented:
@nammbash Your had changes been pulled in. We are having troubles with internal tests again, but I don't think they are related to your PR. I'll let you know if we need more changes here. Sorry for the inconvenience!

penpornk commented:
It's finally merged! Thank everyone so much for the efforts, @nammbash, @rthadur, @annarev, @alextp, and @hgadig!
